### PR TITLE
Workflow steps 2.2, 2.3

### PR DESCRIPTION
**Pull request by mbaynton**
* Allow phpcs warnings to pass the build
* Add some specific exceptions

### CODE REVIEW

File: src/Client/DurableStorage/FilesystemDurableStorage.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+
+namespace Tuf\Client\DurableStorage;
+
+/**
+ * Class FilesystemLocalState
+ *
+ * A simple implementation of LocalStateInterface using the filesystem.
+ * Applications might want to provide an alternative implementation with better performance and error handling.
```
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:10]: We could also add a PHPCS rule from that (borrow Drupal's?).

File: src/Client/DurableStorage/FilesystemDurableStorage.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ */
+class FilesystemDurableStorage implements \ArrayAccess
```
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:12]: @mbaynton clarified that this required #17, which has since been merged. So some of my comments apply to HEAD now as well.

File: src/Client/DurableStorage/FilesystemDurableStorage.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:10]: If we're gonna do it we should add phpcs for it. Can we add that as an additional style-only & associated phpcs rule change to #19 to keep PRs more clean?

[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:10]: +1 to making it a separate PR.

[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:10]: Fixing it in #19

[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:10]: Hmm, having phpcs enforce line length only for comments to 80 doesn't seem to be a "just turn it on" kind of thing. I'm looking in Drupal's Coder module now to see if there's some custom sniff that does it...

PSR-2 recommends 120 chars and is explicit that it is not a hard requirement.

File: src/Client/DurableStorage/FilesystemDurableStorage.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+{
+    /**
+     * @var string $rootPath
+     *   The filesystem directory where durable state is rooted.
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php]: rewrote description and renamed "root" things to "base" things

File: src/Client/DurableStorage/FilesystemDurableStorage.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    public function __construct(string $rootPath)
+    {
+        if (! is_dir($rootPath)) {
+            throw new \RuntimeException("Cannot initialize filesystem local state: \"$rootPath\" is not a directory.");
+        }
+
+        $this->rootPath = $rootPath;
+    }
+
+    protected function offsetToPath($offset)
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:29]: I disagree on this because the \ArrayAccess interface we're implementing uses the term "offset" and the purpose of this method is to translate between array offsets and filenames. Care 2/10.

File: src/Client/DurableStorage/FilesystemDurableStorage.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        return $this->rootPath . DIRECTORY_SEPARATOR . $offset;
+    }
+
+    public function offsetExists($offset)
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:34]: Cannot rename, it's an interface method.

File: src/Client/DurableStorage/FilesystemDurableStorage.php [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    {
+        return file_get_contents($this->offsetToPath($offset));
+    }
+
+    public function offsetSet($offset, $value)
+    {
+        file_put_contents($this->offsetToPath($offset), $value);
+    }
+
+    public function offsetUnset($offset)
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:49]: I'll docblock these as "implements \ArrayAccess::..."

File: src/Client/DurableStorage/FilesystemDurableStorage.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:40]: I'm suggesting renaming the method on the interface as well.

File: src/Client/DurableStorage/FilesystemDurableStorage.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:29]: I mean that we should get rid of the term "offset" entirely because it's confusing/meaningless.

File: src/Client/DurableStorage/ValidatingArrayAccessAdapter.php [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+
+namespace Tuf\Client\DurableStorage;
+
+/**
+ * Class ValidatingArrayAccessAdapter
+ *
+ * Thin wrapper around ArrayAccess inserted between tuf and the backend \ArrayAccess to limit the valid inputs.
+ */
+class ValidatingArrayAccessAdapter implements \ArrayAccess
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/ValidatingArrayAccessAdapter.php:11]: I could get behind `DurableStorageAccessValidator`. It's not meant to be file specific.

Construction of the storage layer, if we stick with an \ArrayAccess-based approach:
* Application provides implementation of \ArrayAccess that might be database, disk, etc
* In php-tuf bootstrapping, construct a `DurableStorageAccessValidator` to wrap the implementation provided by the application
* rest of php-tuf uses array syntax on an instance of `DurableStorageAccessValidator`

File: src/Client/Updater.php [hunk 8]
```diff
@@ -45,25 +49,14 @@ public function __construct($repository_name, $mirrors)
   /**
    * @todo Update from python comment https://github.com/theupdateframework/tuf/blob/1cf085a360aaad739e1cc62fa19a2ece270bb693/tuf/client/updater.py#L999
    *
-   * @param bool $unsafely_update_root_if_necessary
    */
-    public function refresh($unsafely_update_root_if_necessary = true)
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:50]: I think we'll bring it back when we get a little further along. Only deleted based on feeling that this group seems to want to keep unused stuff out of the code.

File: src/Client/Updater.php [hunk 9]
```diff
@@ -45,25 +49,14 @@ public function __construct($repository_name, $mirrors)
...
-   * @param $target_stream
-   *
-   * @return bool
-   *   Returns true if the target validates.
-   */
-    public function validateTarget($target_repo_path, $target_stream)
-    {
-        // @TODO source original "step 0" root data from client state, not repository
-        $root_data = json_decode($this->getRepoFile('root.json'), true);
+        // @TODO Inject DurableStorage
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:55]: >  How do we want to handle service injection?
As simply as possible :stuck_out_tongue: 

So far this is the only "injected" thing that we're planning on having the app provide to us. I think we should just ask for it in the `Updater` class's constructor and that's it.

In general I think hitching our wagon to a specific DI implementation in a library like this isn't a good idea because it couples it to a specific subset of the php ecosystem, like Symfony.

File: src/Client/Updater.php [hunk 10]
```diff
@@ -45,25 +49,14 @@ public function __construct($repository_name, $mirrors)
...
+        $durableStorage = new ValidatingArrayAccessAdapter(
+            new FilesystemDurableStorage(__DIR__ . "/../../fixtures/tufrepo/metadata")
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:57]: Yes.

File: src/Client/Updater.php [hunk 11]
```diff
@@ -105,13 +98,77 @@ public function validateTarget($target_repo_path, $target_stream)
...
+        $this->checkRollbackAttack($currentStateTimestamp['signed'], $timestamp_structure['signed']);
+
+        // SPEC: 2.3
+        $this->checkFreezeAttack($timestamp_structure['signed'], $now_date);
+        $durableStorage['timestamp.json'] = $timestamp_contents;
+
         return true;
     }
 
+    protected function metadataTimestampToDatetime(string $timestamp) : \DateTimeImmutable
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:115]: We should add them everywhere.

Chatted with Ted and others yesterday about this. It sounds like we'd be able to automatically remove these.
https://drupal.slack.com/archives/C7QJNEY3E/p1596747106078100

File: src/Client/Updater.php [hunk 12]
```diff
@@ -161,7 +218,7 @@ private function getRepoFile($string)
     {
         try {
           // @todo Ensure the file does not exceed a certain size to prevent DOS attacks.
-            return file_get_contents(__DIR__ .  "/../../fixtures/tufclient/tufrepo/metadata/current/$string");
+            return file_get_contents(__DIR__ .  "/../../fixtures/tufrepo/metadata/$string");
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:221]: Hardcoding not added by this PR. We obviously need to fix it as we build the project out.

File: src/Exception/NotFoundException.php [hunk 13]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+
+namespace Tuf\Exception;
+
+use Throwable;
+
+class NotFoundException extends \Exception
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Exception/NotFoundException.php:8]: Good idea. Although the jokes about tough exceptions pretty much write themselves...

File: src/Client/Updater.php [hunk 14]
```diff
@@ -79,20 +72,20 @@ public function validateTarget($target_repo_path, $target_stream)
...
 
         // SPEC: 1.8.
         $expires = $signed['expires'];
         $fake_now = '2020-08-04T02:58:56Z';
-        $expire_date = \DateTime::createFromFormat("Y-m-dTH:i:sZ", $fake_now);
-        $now_date = \DateTime::createFromFormat("Y-m-dTH:i:sZ", $expires);
+        $expire_date = $this->metadataTimestampToDatetime($expires);
+        $now_date = $this->metadataTimestampToDatetime($fake_now);
         if ($now_date > $expire_date) {
             throw new \Exception("Root has expired. Potential freeze attack!");
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:86]: Sure. Using Python TUF as a reference for which types of exceptions we'll ultimately want, I'm adding a `FormatException` and throwing it here.

File: src/Exception/NotFoundException.php [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    public $key;
+
+    /**
+     * @var string $itemType
+     *   The type of item (signing key, file, etc.) that was not found.
+     */
+    public $itemType;
+
+    public function __construct($key = "", $itemType = "item", $code = 0, Throwable $previous = null)
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Exception/NotFoundException.php:22]: Fair I suppose if we're not explicitly using it in a subclass we should just default it to 0 when calling \Exception::__construct

File: tests/Client/UpdaterTest.php [hunk 16]
```diff
@@ -7,29 +7,45 @@
...
+    protected function getSystemInTest()
     {
         $mirrors = array(
-        'mirror1' => array(
-        'url_prefix' => 'http://localhost:8001',
-        'metadata_path' => 'metadata',
-        'targets_path' => 'targets',
-        'confined_target_dirs' => array(),
-        ),
+            'mirror1' => array(
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply tests/Client/UpdaterTest.php:16]: I see you filed an issue for it :smiley:

File: src/Client/DurableStorage/FilesystemDurableStorage.php [hunk 17]
```diff
@@ -0,0 +1,53 @@
+<?php
+
+
+namespace Tuf\Client\DurableStorage;
```
[the remark this answers is not in the document]
Comment (davidstrauss) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:3]: +1

[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/DurableStorage/FilesystemDurableStorage.php:3]: Oh are we talking about renaming the class, not the namespace? The way github presents this comment I read it as a suggestion to rename the namespace.

File: src/Client/DurableStorage/FilesystemDurableStorage.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/DurableStorage/FilesystemDurableStorage.php]: Discussed with @mbaynton that these are the `\ArrayAccess` methods, not a new public interface. There is no public interface currently. Maybe there should be?

File: src/Client/Updater.php (hunk 8 above)
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/Updater.php]: I added the `@todo` and removed the unused flag in a separate commit, so this is unneeded now.

File: .github/workflows/php.yml [hunk 18]
```diff
@@ -58,7 +58,7 @@ jobs:
         run: composer lint
 
       - name: Run PHPCS
-        run: composer phpcs
+        run: composer phpcs-ci
```
Comment (xjm) [inline .github/workflows/php.yml:61]: What does this do? Is it in scope here?

File: src/Client/DurableStorage/DurableStorageAccessValidator.php [hunk 19]
```diff
@@ -4,11 +4,11 @@
 namespace Tuf\Client\DurableStorage;
 
 /**
- * Class ValidatingArrayAccessAdapter
+ * Class DurableStorageAccessValidator
  *
  * Thin wrapper around ArrayAccess inserted between tuf and the backend \ArrayAccess to limit the valid inputs.
  */
-class ValidatingArrayAccessAdapter implements \ArrayAccess
+class DurableStorageAccessValidator implements \ArrayAccess
```
Comment (xjm) [inline src/Client/DurableStorage/DurableStorageAccessValidator.php:11]: 👍 Much better name.

File: src/Client/DurableStorage/FileStorage.php [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+            throw new \RuntimeException("Cannot initialize filesystem local state: \"$basePath\" is not a directory.");
+        }
+
+        $this->basePath = $basePath;
+    }
+
+    protected function offsetToPath($offset)
+    {
+        return $this->basePath . DIRECTORY_SEPARATOR . $offset;
+    }
```
Comment (xjm) [inline src/Client/DurableStorage/FileStorage.php:34]: So this helper method is protected (which I missed before) and is the only method not required to implement `\ArrayAccess`.

It needs a docblock. I also still think we need a different method name because I don't know what it means from reading the name. Maybe just `pathWithBasePath()` or something?

File: src/Client/Updater.php [hunk 21]
```diff
@@ -52,23 +55,10 @@ public function __construct($repositoryName, $mirrors)
...
-    }
-
-  /**
-   * Validates a target.
-   *
-   * @param $targetRepoPath
-   * @param $targetStream
-   *
-   * @return bool
-   *   Returns true if the target validates.
```
Comment (xjm) [inline src/Client/Updater.php:65]: Hm, why are we deleting this?

Or is this just a diff artifact?

File: src/Client/Updater.php [hunk 22]
```diff
@@ -112,13 +102,81 @@ public function validateTarget($targetRepoPath, $targetStream)
...
+    /**
+     * Verifies that an incoming remote version of a metadata file is >= the last known version.
+     *
+     * @param $localMetadata
+     * @param $remoteMetadata
+     * @throws RollbackAttackException
+     */
+    protected function checkRollbackAttack(array $localMetadata, array $remoteMetadata)
+    {
+        $localVersion = $localMetadata['version'] + 0;
```
Comment (xjm) [inline src/Client/Updater.php:137]: I agree with Ted that this is a rather opaque line. If the intent is to cast it to an int, why not just `(int)`?

File: src/Exception/NotFoundException.php [hunk 23]
```diff
@@ -8,7 +8,7 @@
 /**
  * @TODO Remove this class if not used.
  */
-class NotFoundException extends \Exception
+class NotFoundException extends TufException
```
Comment (xjm) [inline src/Exception/NotFoundException.php:11]: 👍

File: src/Client/Updater.php [hunk 24]
```diff
@@ -86,20 +76,20 @@ public function validateTarget($targetRepoPath, $targetStream)
         $nextVersion = $version + 1;
         $nextRootContents = $this->getRepoFile("$nextVersion.root.json");
         if ($nextRootContents) {
-          // @todo 🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥Add steps do root rotation spec steps 1.3 -> 1.7.
-          //  Not production ready🙀.
+            // @todo 🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥🔥Add steps do root rotation spec steps 1.3 -> 1.7.
+            //  Not production ready🙀.
```
Comment (xjm) [inline src/Client/Updater.php:80]: All these whitespace fixes aren't really in scope; should we fix them in a separate commit?

Comment (xjm) [reply src/Client/Updater.php:80]: Also why is phpcs not failing locally for me if these are indented wrong? 🤔

File: tests/Client/UpdaterTest.php [hunk 25]
```diff
@@ -7,29 +7,45 @@
...
-        'metadata_path' => 'metadata',
-        'targets_path' => 'targets',
-        'confined_target_dirs' => [],
-        ],
+            'mirror1' => [
+                'url_prefix' => 'http://localhost:8001',
+                'metadata_path' => 'metadata',
+                'targets_path' => 'targets',
+                'confined_target_dirs' => [],
+            ],
```
Comment (xjm) [inline tests/Client/UpdaterTest.php:21]: Ditto about whitespace/scope.

**Review by xjm — DISMISSED** [review]

File: .github/workflows/php.yml (hunk 18 above)
Comment (mbaynton) [reply .github/workflows/php.yml:61]: It does this https://drupal.slack.com/archives/C7QJNEY3E/p1596825151112600
> allowing the CI to pass if phpcs shows only warnings, no errors

Hazards of calls / verbal communication, there's not a lot of written record for +1s. @heddn brought it up right as he headed out and I think he understood the proposal and was in favor.

It'll be a lot of shuffling around to split it into another PR and then merge it back into this branch so that the CI passes, but if there's an important reason to...

File: src/Client/Updater.php (hunk 21 above)
Comment (mbaynton) [reply src/Client/Updater.php:65]: All the things we've done so far do not happen when the application validates a single target, but rather when the application checks the repository for updates. The Python library has a `refresh()` function that applications call to begin an update check. So this is just moving the functionality to the right place, inside our `refresh()`

Now `refresh()` isnt empty and `validateTarget` is. The diff is weird.

File: src/Client/Updater.php (hunk 22 above)
Comment (mbaynton) [reply src/Client/Updater.php:137]: Muscle memory when casting strings to numeric in PHP, I usually use this way because it universally works for ints and floats. TUF constrains us to ints here so yeah we can use `(int)`

File: src/Client/Updater.php (hunk 24 above)
Comment (mbaynton) [reply src/Client/Updater.php:80]: Do you disagree with the change?

I rarely object to the exact way a non-controversial change gets done.

Comment (xjm) [reply src/Client/Updater.php:80]: The change itself is correct, but mixing in out-of-scope CS standards makes review harder. If this were a core issue I would request that the CS fixes be done in a separate patch. https://www.drupal.org/core/scope#creep

I'm also wondering why phpcs isn't failing on HEAD when the indentation is wrong like this. Any ideas?

File: .github/workflows/php.yml (hunk 18 above)
Comment (xjm) [reply .github/workflows/php.yml:61]: I would prefer hard fails for maintainability. What in this PR causes it to not pass? Is there a reason not to fix whatever's failing?

File: tests/Client/UpdaterTest.php [hunk 26]
```diff
@@ -7,29 +7,45 @@
...
+
+    /*
+    public function testValidateTarget()
     {
+        $updater = $this->getSystemInTest();
+        $fixture_target = 'testtarget.txt';
+        $target_stream = fopen('data://text/plain,' . 'Test File', 'r');
+        $this->assertTrue($updater->validateTarget($fixture_target, $target_stream));
     }
+    */
```
Comment (xjm) [inline tests/Client/UpdaterTest.php:50]: Just noticed that this is commented out. Why?

Comment (mbaynton) [reply tests/Client/UpdaterTest.php:50]: The newly emptied function it tests does not return true and failing tests is a bad look.

Comment (mbaynton) [reply tests/Client/UpdaterTest.php:50]: (Coverage is maintained by the new `testRefreshRepository()`)

Comment (xjm) [reply tests/Client/UpdaterTest.php:50]: Can we file an issue for it then and add a `@todo` comment that links said issue? Commented code is dead code.

Comment (mbaynton) [reply tests/Client/UpdaterTest.php:50]: Can we just delete it for now? We won't forget to implement validating targets if we don't file an issue :wink:

Comment (xjm) [reply tests/Client/UpdaterTest.php:50]: That's true; it's kind of critical to the functionality. 😛  I'm fine with just deleting it for now. I just feel very strongly about not having commented code or `@todo`s without documented issues for them. 🙂

Comment (mbaynton) [reply tests/Client/UpdaterTest.php:50]: >  or @todos without documented issues for them.

Good to know

Comment (mbaynton) [reply tests/Client/UpdaterTest.php:50]:
```suggestion
```

File: .github/workflows/php.yml (hunk 18 above)
Comment (mbaynton) [reply .github/workflows/php.yml:61]: Since we don't all agree about this change, there is a reason for it to be looked at as a separate PR. I will pull it out.

Comment (mbaynton) [reply .github/workflows/php.yml]: > What in this PR causes it to not pass? 

Three lines were too long. I've changed things around so it works for now.

> Is there a reason not to fix whatever's failing?

See #26

File: src/Client/Updater.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php]: Can this be resolved? I don't think a goal of this PR, which was supposed to just add two more steps to the workflow, should be to transform all the day-1 happy path code into release-quality code.

File: src/Client/Updater.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php]: Fixed by #24

**Comment by mbaynton** [comment]
I've looked over all the not resolved comments again. Lots are related to the word "offset" that we can't get rid of without doing storage differently, some are related to changes to whitespace that we agree on anyway, and others I think are just a product of the diffs having to touch code that was already and still is quick & dirty.

@xjm if there's anything substantial left that I'm not addressing from the 70+ comments, can you summarize those concerns again or approve it?

File: src/Client/Updater.php (hunk 24 above)
Comment (xjm) [reply src/Client/Updater.php]: I investigated further, then fixed this in `main` and filed an issue to get PHPCS working for these indentation rules:
https://github.com/php-tuf/php-tuf/issues/30

**Comment by xjm** [comment]
#31 and #32 split out some of the changes from this PR in order to reduce what needs to be reviewed here.

File: src/Client/Updater.php [hunk 27]
```diff
@@ -105,13 +98,77 @@ public function validateTarget($target_repo_path, $target_stream)
...
+   *
+   * @param $target_repo_path
+   * @param $target_stream
+   *
+   * @return bool
+   *   Returns true if the target validates.
+   */
+    public function validateTarget($target_repo_path, $target_stream)
+    {
+    }
```
[the remark this answers is not in the document]
Comment (mbaynton) [reply src/Client/Updater.php:202]: I thought we chatted about this and agreed the todo wasn't necessary here since we wouldn't forget but I see this is still open so I created an issue. Perhaps I misunderstood the conclusion of that conversation. Resolving.

File: src/Client/Updater.php [hunk 28]
```diff
@@ -129,14 +119,99 @@ public function validateTarget($targetRepoPath, $targetStream)
...
             throw new \Exception("Improperly signed repository timestamp.");
         }
 
+        // SPEC: 2.2
+        $currentStateTimestamp = json_decode($this->durableStorage['timestamp.json'], true);
+        $this->checkRollbackAttack($currentStateTimestamp['signed'], $timestampStructure['signed']);
+
+        // SPEC: 2.3
+        $this->checkFreezeAttack($timestampStructure['signed'], $nowDate);
+        $durableStorage['timestamp.json'] = $timestampContents;
```
Comment (xjm) [inline src/Client/Updater.php:131]: `checkSignatures()` returns a bool and then `refresh()` throws an exception if it's wrong. However, the new checks return `void` and throw the exceptions themselves. We should make the behavior consistent. Thoughts on which is best? 

Returning booleans might give us more flexibility for testing and reuse, but OTOH we can always `try/catch`. Always throwing the exceptions within the checkers might be safer, because it ensures a caller can't accidentally introduce a bypass of the check. What do you think?

File: src/Client/DurableStorage/FilesystemDurableStorage.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/DurableStorage/FilesystemDurableStorage.php]: This is fixed now in the branch and we have #26 to discuss the overall question of comment line length warnings, so marking resolved since the rest is out of scope. Thanks!

**Comment by xjm** [comment]
I split out #37 as well.

File: src/Client/Updater.php (hunk 28 above)
Comment (tedbow) [reply src/Client/Updater.php:131]: I like the idea of throwing exceptions

File: src/Client/Updater.php [hunk 29]
```diff
@@ -3,8 +3,11 @@
 
 namespace Tuf\Client;
 
-use Tuf\Client\DurableStorage\FilesystemDurableStorage;
+use Tuf\Client\DurableStorage\FileStorage;
```
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/Updater.php:6]: Ah, I was wondering why this bug wasn't causing a CI fail..

File: src/Client/Updater.php (hunk 28 above)
Comment (xjm) [reply src/Client/Updater.php]: We're thowing exceptions in any case.. it's a question as to where in the callstacks the exception is thrown: within the checker, or in response to the checker's result. I assming Ted means "Exceptions thrown  by the checker,  not the caller."

Comment (xjm) [reply src/Client/Updater.php]: @tedbow Can you clarify what you meant?

**Comment by xjm** [comment]
@tedbow Throwing exceptions within the checkers, or in response to the checkers' results? See https://github.com/php-tuf/php-tuf/pull/18/files/3d5991f33483c76c6050c9eebd84f7554c3c6f15#r468512490

File: src/Client/Updater.php (hunk 29 above)
[the remark this answers is not in the document]
Comment (xjm) [reply src/Client/Updater.php]: The unused use rule is another one that's in the Drupal ruleset. There's an open PR to add it to the generic ruleset, but it's not yet merged: https://github.com/squizlabs/PHP_CodeSniffer/pull/1106

Meanwhile, I pushed https://github.com/php-tuf/php-tuf/pull/18/commits/a4d9813f27ec5f0653632a1f754cb6061d626406 to resolve this one.

File: src/Client/Updater.php (hunk 28 above)
Comment (tedbow) [reply src/Client/Updater.php]: I meant inside the method not the caller.

File: src/Client/Updater.php [hunk 30]
```diff
@@ -129,14 +119,99 @@ public function validateTarget($targetRepoPath, $targetStream)
...
+     *     Thrown if a potential rollback attack is detected.
+     */
+    protected function checkRollbackAttack(array $localMetadata, array $remoteMetadata)
+    {
+        $localVersion = (int) $localMetadata['version'];
+        if ($localVersion == 0) {
+            // Failsafe: if local metadata just doesn't have a version property or it is not an integer,
+            // we can't perform this check properly.
+            $message = "Empty or invalid local timestamp version \"${localMetadata['version']}\"";
+            throw new RollbackAttackException($message);
```
[the remark this answers is not in the document]
Comment (tedbow) [reply src/Client/Updater.php:187]: I think this is fine.

File: src/Client/Updater.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (davidstrauss) [reply src/Client/Updater.php]: I've been assuming that we could have the caller handle injection/services however they'd like by supply an implementation during construction. I'd prefer to not adopt a specific services configuration pattern within this library itself.

File: src/Client/Updater.php (hunk 28 above)
Comment (tedbow) [reply src/Client/Updater.php]: Changed \Tuf\Client\Updater::checkSignatures to throw an exceptions like the other check methods

File: src/Client/Updater.php [hunk 31]
```diff
@@ -129,14 +119,99 @@ public function validateTarget($targetRepoPath, $targetStream)
...
+        }
+        $remoteVersion = (int) $remoteMetadata['version'];
+        if ($remoteVersion < $localVersion) {
+            $message = "Remote timestamp metadata version \"${remoteMetadata['version']}\"" .
+                " is less than previously seen timestamp version \"${localMetadata['version']}\"";
```
[the remark this answers is not in the document]
Comment (tedbow) [reply src/Client/Updater.php]: Fixed this

**Review by tedbow — APPROVED** [review]
